Add tutorial on how to supplement rosdep with custom keys (#6187)
ros2/ros2_documentation: source/Tutorials/Intermediate/Supplementing-Custom-Rosdep-Keys.rst

.. code-block:: console

   $ rosdep where-defined libopencv-dev
   https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml
   $ rosdep resolve libopencv-dev
   #apt
   libopencv-dev

In summary, it allows ``rosdep`` to resolve the ``libopencv-dev`` key to the ``apt`` package of the same name.

Note that, from the output above, we can deduce that the command was run on a Ubuntu or Debian OS.
On RHEL, the key instead resolves to the DNF package ``opencv-devel``:

.. code-block:: console

   $ rosdep resolve libopencv-dev --os=rhel:9
   #dnf
   opencv-devel

Extending ``rosdep`` with a custom sources file
-----------------------------------------------

.. This sheds light on what we need to do to get ``rosdep`` to understand new keys: add a new custom sources file!

The above hopefully makes it clear what needs to be done to get ``rosdep`` to understand new keys: add a new custom sources file!

As a toy example, let's add a new sources file telling ``rosdep`` to fetch keys from a YAML file stored on the local machine.

> Fire up your favorite text editor and write the following into ``/etc/ros/rosdep/sources.list.d/10-custom.list``:

Reviewer: I would recommend using a higher number by default with a comment that you should use a lower number if you want to override. 

But it will be less likely to cause problems if the default is to extend for undefined rules. 
```suggestion
Fire up your favorite text editor and write the following into ``/etc/ros/rosdep/sources.list.d/30-custom.list``:
```

.. code-block:: yaml

  yaml file:///etc/ros/rosdep/custom_rules.yaml

Now write the following into ``/etc/ros/rosdep/custom_rules.yaml``:

.. code-block:: yaml

  awesome_library:
    ubuntu: [awesome_library]
  that_other_library:
    ubuntu:
      pip:
        packages: [another_library]

This defines two new rosdep rules:

1. The key ``awesome_library``, defined only for Ubuntu, mapping to the ``apt`` package of the same name
2. The key ``that_other_library``, defined only for Ubuntu, mapping to the ``pip`` package named ``another_library``

> After running ``rosdep update``, ``rosdep`` will detect the new ``10-custom.list``, prompting it to scan the contents of the ``custom_rules.yaml`` file.

Reviewer: ```suggestion
After running ``rosdep update``, ``rosdep`` will detect the new ``30-custom.list``, prompting it to scan the contents of the ``custom_rules.yaml`` file.
```
Matching change to suggestion above

Now ``rosdep`` is set up to recognize these new keys and what they should map to:

.. code-block:: console

   $ rosdep resolve awesome_library
   #apt
   awesome_library
   $ rosdep resolve that_other_library
   #pip
   another_library

Now all you have to do is add ``<depend>awesome_library</depend>`` to your ROS packages ``package.xml``, and ``rosdep`` will know how to install the dependency!

Closing remarks
---------------

The toy example above only hints at what is possible with custom rosdep keys.

- **Is your dependency a apt package hosted in a third party PPA?**
  Not a problem.
  Since all ``rosdep`` does is converting the key to a ``apt install`` invocation, ``apt`` will have no problem installing the package (provided you have added the PPA).
- **Is your dependency a pip package hosted in a third party index?**
  Add the index to your ``pip.conf`` and you're good to go.
- **Sources files don't have to point to files on the local machine.**
  Both ``file://`` and ``https://`` syntax is supported.

> 

Reviewer: ```suggestion
- **Sources are loaded in alphabetical order** 
  If you add a conflicting rule in the 30 prefix it will not be used.
  If you created a sources file with a 10 prefix it will override packages in the default list (prefix 20).
  If you're using anything from binaries it's highly recommended not to override dependency declarations as it will likely cause binary incompatibilities which can be very hard to  debug. 
```

Calling out the ordering

This is the preferred course of action when you have some dependency (like a ``apt`` or ``pip`` package) that you wish to be able to install via ``rosdep``.

However, there are many cases when contributing your keys directly might be difficult.
For example, if the dependency

1. is not available on the target distributions default APT (or pip) repositories
2. is a proprietary library
3. is some niche library which is useful only for you or your organization
4. is a ROS package you :doc:`built and packaged yourself <../../How-To-Guides/Building-a-Custom-Deb-Package>`, but don't wish to share with the broader ROS community

While the option exists to :doc:`fork rosdistro <../../How-To-Guides/Using-Custom-Rosdistro>` in its entirety, this might be overkill if you just want to keep using a official ROS distribution as usual, only with some extra rosdep keys defined on top.
This tutorial explains how to achieve this.

Preliminaries: How ``rosdep`` fetches rosdep keys
-------------------------------------------------

> .. Let's first review some details of ``rosdep`` which will be relevant to the remainder of this tutorial.

Reviewer 2: Are these commented-out sections needed in the rst?

Author: Ah thanks, these are leftover alternative formulations for the first sentence that I was contemplating. Fixed in ad8bbe8f876fe38626e571cb98134ad04dcf8a02.

.. In order to get a good understanding of what we're about to do, let's first explore how ``rosdep`` works from first principles.

In order to get a good understanding of what we're about to do, let's first explore some relevant details about how ``rosdep`` works.

``rosdep`` is similar to other tools like ``apt`` that use a sources list to maintain a local index.
These sources are stored in ``/etc/ros/rosdep/sources.list.d``.

> This is identical to how apt stores repositories in ``/etc/apt/sources.list.d``.

Reviewer 2: ```suggestion
This is similar to how apt stores repositories in ``/etc/apt/sources.list.d``.
```
It is not identical, APT uses a different format

By default (as part of first-time setup, ``rosdep init``), you only have a single sources file: ``/etc/ros/rosdep/sources.list.d/20-default.list``.
Inspecting its contents, you will see entries like these:

.. code-block:: console

   $ cat /etc/ros/rosdep/sources.list.d/20-default.list
   ...
   yaml https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml
   yaml https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml
   ...

These entries are what dictates where ``rosdep`` fetches rosdep keys and their mappings (rosdep **rules**) from when ``rosdep update`` is called.
When called, ``rosdep`` compiles the relevant contents from all declared entries in all sources files into a local cached index.
This local index is then used when installing or looking up ("resolving") rosdep keys.

> For example, the fact that the first entry defines the ``libopencv-dev`` key (see `here <https://github.com/ros/rosdistro/blob/72f24d6/rosdep/base.yaml#L5240-L5252>`_) is what allows ``rosdep`` to resolve it:

Reviewer 2: ```suggestion
For example, the fact that the first entry (``base.yaml``) defines the ``libopencv-dev`` key (see `here <https://github.com/ros/rosdistro/blob/72f24d6/rosdep/base.yaml#L5240-L5252>`_) is what allows ``rosdep`` to resolve it:
```

.. code-block:: console

   $ rosdep where-defined libopencv-dev
   https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml
   $ rosdep resolve libopencv-dev
   #apt
   libopencv-dev

In summary, it allows ``rosdep`` to resolve the ``libopencv-dev`` key to the ``apt`` package of the same name.

Note that, from the output above, we can deduce that the command was run on a Ubuntu or Debian OS.
On RHEL, the key instead resolves to the DNF package ``opencv-devel``:

.. code-block:: console

   $ rosdep resolve libopencv-dev --os=rhel:9
   #dnf
   opencv-devel

Extending ``rosdep`` with a custom sources file
-----------------------------------------------

.. This sheds light on what we need to do to get ``rosdep`` to understand new keys: add a new custom sources file!

The above hopefully makes it clear what needs to be done to get ``rosdep`` to understand new keys: add a new custom sources file!

As a toy example, let's add a new sources file telling ``rosdep`` to fetch keys from a YAML file stored on the local machine.

> Fire up your favorite text editor and write the following into ``/etc/ros/rosdep/sources.list.d/30-custom.list``:

Reviewer 2: ```suggestion
Fire up your favorite text editor and write the following into ``/etc/ros/rosdep/sources.list.d/30-custom.list`` (the editor will need to be launched with root privileges, e.g. via ``sudo``):
```

  yaml file:///etc/ros/rosdep/custom_rules.yaml

Now write the following into ``/etc/ros/rosdep/custom_rules.yaml``:

.. code-block:: yaml

  awesome_library:
    ubuntu: [awesome_library]
  that_other_library:
    ubuntu:
      pip:
        packages: [another_library]

This defines two new rosdep rules:

1. The key ``awesome_library``, defined only for Ubuntu, mapping to the ``apt`` package of the same name
2. The key ``that_other_library``, defined only for Ubuntu, mapping to the ``pip`` package named ``another_library``

After running ``rosdep update``, ``rosdep`` will detect the new ``30-custom.list``, prompting it to scan the contents of the ``custom_rules.yaml`` file.
Now ``rosdep`` is set up to recognize these new keys and what they should map to:

.. code-block:: console

   $ rosdep resolve awesome_library
   #apt
   awesome_library
   $ rosdep resolve that_other_library
   #pip
   another_library

Now all you have to do is add ``<depend>awesome_library</depend>`` to your ROS packages ``package.xml``, and ``rosdep`` will know how to install the dependency!

Closing remarks
---------------

The toy example above only hints at what is possible with custom rosdep keys.

> - **Is your dependency a apt package hosted in a third party PPA?**

Reviewer 2: ```suggestion
- **Is your dependency an APT package hosted in a third party PPA?**
```

  Not a problem.

>   Since all ``rosdep`` does is converting the key to a ``apt install`` invocation, ``apt`` will have no problem installing the package (provided you have added the PPA).

Reviewer 2: ```suggestion
  Since all ``rosdep`` does is converting the key to a ``apt install`` invocation, APT will have no problem installing the package (provided you have added the PPA).
```

- **Is your dependency a pip package hosted in a third party index?**
  Add the index to your ``pip.conf`` and you're good to go.
- **Sources files don't have to point to files on the local machine.**

>   Both ``file://`` and ``https://`` syntax is supported.

Reviewer 2: ```suggestion
  Both ``file://`` and ``https://`` syntax is supported (on Linux, absolute paths start with ``/``, which leads to a triple slash like ``file:///etc/rosdep/my.file``).
```

- **Sources are loaded in alphabetical order.**
  If you add a conflicting rule in the 30 prefix it will not be used.
  If you created a sources file with a 10 prefix it will override packages in the default list (prefix 20).

>   If you're using anything from binaries it's highly recommended not to override dependency declarations as it will likely cause binary incompatibilities which can be very hard to debug.

Reviewer 2: ```suggestion
  If you're using packages installed from binary repositories it's highly recommended to not override dependency declarations as it will likely cause binary incompatibilities which can be very hard to debug.
```
